components/browser/thumbnails/src/main/java/mozilla/components/browser/thumbnails/storage/ThumbnailStorage.kt

```diff
@@ -0,0 +1,79 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package mozilla.components.browser.thumbnails.storage
+
+import android.content.Context
+import android.graphics.Bitmap
+import androidx.annotation.WorkerThread
+import kotlinx.coroutines.CoroutineDispatcher
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Deferred
+import kotlinx.coroutines.asCoroutineDispatcher
+import kotlinx.coroutines.async
+import mozilla.components.browser.thumbnails.R
+import mozilla.components.browser.thumbnails.utils.ThumbnailDiskCache
+import mozilla.components.support.base.log.logger.Logger
+import mozilla.components.support.images.DesiredSize
+import mozilla.components.support.images.decoder.AndroidImageDecoder
+import java.util.concurrent.Executors
+
+private const val MAXIMUM_SCALE_FACTOR = 2.0f
+
+// Number of worker threads we are using internally.
+private const val THREADS = 3
+
+internal val sharedDiskCache = ThumbnailDiskCache()
+
+/**
+ * Thumbnail storage layer which handles saving and loading the thumbnail from the disk cache.
+ */
+class ThumbnailStorage(
+    private val context: Context,
+    jobDispatcher: CoroutineDispatcher = Executors.newFixedThreadPool(THREADS).asCoroutineDispatcher()
+) {
+    private val decoders = AndroidImageDecoder()
+    private val logger = Logger("ThumbnailStorage")
+    private val maximumSize = context.resources.getDimensionPixelSize(R.dimen.mozac_browser_thumbnails_maximum_size)
+    private val scope = CoroutineScope(jobDispatcher)
+
+    /**
+     * Asynchronously loads a thumbnail [Bitmap] for the given session ID or url.
+     */
+    fun loadThumbnail(sessionIdOrUrl: String): Deferred<Bitmap?> = scope.async {
+        loadThumbnailInternal(sessionIdOrUrl).also { loadedThumbnail ->
+            if (loadedThumbnail != null) {
+                logger.debug("Loaded thumbnail from disk (sessionIdOrUrl = $sessionIdOrUrl)")
+            } else {
+                logger.debug("No thumbnail loaded (sessionIdOrUrl = $sessionIdOrUrl)")
+            }
+        }
+    }
+
+    @WorkerThread
+    private fun loadThumbnailInternal(sessionIdOrUrl: String): Bitmap? {
+        val desiredSize = DesiredSize(
+            targetSize = context.resources.getDimensionPixelSize(R.dimen.mozac_browser_thumbnails_size_default),
+            maxSize = maximumSize,
+            maxScaleFactor = MAXIMUM_SCALE_FACTOR
```

Comment:
I noticed the icons are noticeably blurry when comparing from the cache to the ones fresh in-memory. Maybe we need to change the scaling value?

<img width="523" alt="Screen Shot 2020-05-27 at 1 30 10 PM" src="https://user-images.githubusercontent.com/1370580/83053134-411b6980-a01e-11ea-90ee-7012895de004.png">

Reply:
Gonna address this as a followup.